contracts/Redemption.vy

```diff
@@ -0,0 +1,502 @@
+    CURVE_POOL = CurvePoolInterface(curve_pool)
+    self._transfer_ownership(owner)
+    self.payee = owner
+    self.packed_scaling_factor = shift(scaling_factor, 128) | shift(scaling_factor, 192)
+
+
+@payable
+@external
+def redeem(amount: uint256, recipient: address = msg.sender) -> uint256:
+    """
+    @notice Redeem your dYFI for YFI using ETH.
+    @dev Redemption tolerates a 0.3% negative or positive slippage.
+    @param amount amount of dYFI to spend
+    @param recipient of the exercised YFI
+    """
+    self._check_killed()
+    assert YFI.balanceOf(self) >= amount, "not enough YFI"
+    eth_required: uint256 = self._eth_required(amount)
+    assert eth_required > 0
+    tolerance: uint256 = eth_required * SLIPPAGE_TOLERANCE / SLIPPAGE_DENOMINATOR
+    if msg.value < (eth_required - tolerance) or msg.value > (eth_required + tolerance):
+        raise "price out of tolerance"
+    DYFI.burn(msg.sender, amount)
+    raw_call(self.payee, b"", value=msg.value)
+    YFI.transfer(recipient, amount)
+    return amount
+
+
+@external
+@view
+def discount() -> uint256:
+    """
+    @notice Get the current dYFI redemption discount
+    @return Redemption discount (18 decimals)
+    @dev 
+        Discount formula is `1/(1 + 10 e^(4.7(s*x - 1)))`,
+        with `x = veyfi supply / yfi supply`
+    """
+    return self._discount()
+
+
+@internal
+@view
+def _discount() -> uint256:
+    yfi_supply: uint256 = YFI.totalSupply()
+    veyfi_supply: uint256 = VEYFI.totalSupply()
+    x: int256 = convert(veyfi_supply * UNIT / yfi_supply, int256)
+    x = self._exp(47 * (self._scaling_factor()[0] * x / E18 - E18) / 10)
+    return convert(E18 * E18 / (E18 + 10 * x), uint256)
+
+
+@external
+@view
+def eth_required(amount: uint256) -> uint256:
+    """
+    @notice Estimate the required amount of ETH to redeem the amount of dYFI for YFI
+    @param amount Amount of dYFI
+    @return Amount of ETH required
+    """
+    return self._eth_required(amount)
+
+
+@internal
+@view
+def _eth_required(amount: uint256) -> uint256:
+    return amount * self._get_latest_price() / UNIT * (UNIT - self._discount()) / UNIT
+
+
+@external
+@view
+def get_latest_price() -> uint256:
+    """
+    @notice Get the latest price of YFI in ETH
+    @return Price of YFI in ETH (18 decimals)
+    """
+    return self._get_latest_price()
+
+
+@internal
+@view
+def _get_latest_price() -> uint256:
+    oracle_price: uint256 = convert(self._get_oracle_price(), uint256)
+    pool_price: uint256 = CURVE_POOL.price_oracle()
+    if pool_price < oracle_price:
+        return oracle_price
+    return pool_price
+
+
+@internal
+@view
+def _get_oracle_price() -> int256:
+    round_id: uint80 = 0
+    price: int256 = 0
+    started_at: uint256 = 0
+    updated_at: uint256 = 0
+    answered_in_round: uint80 = 0
+    (round_id, price, started_at, updated_at, answered_in_round) = PRICE_FEED.latestRoundData()
+    assert updated_at + 86400 > block.timestamp, "price too old"
```

Comment:
https://docs.chain.link/data-feeds/price-feeds/addresses/?network=ethereum&page=1&categories=verified&search=0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419

https://docs.chain.link/data-feeds/price-feeds/addresses/?network=ethereum&page=1&categories=verified&search=0xA027702dbb89fbd58938e4324ac03B58d812b0E1

according to these pages the heartbeat shouldnt be higher than 3600s from each of those sources, so this 24 hour gap may be reduced considerably given we are using combining sources now?